File: Assets/_IntelliVerseXSDK/Multiplayer/Core/IVXGameModeManager.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace IntelliVerseX.GameModes
{
    /// <summary>
    /// Central singleton for game mode selection, player slot management, and match lifecycle.
    /// Attach to a persistent GameObject or let it auto-create via Instance.
    /// </summary>
    [HelpURL("https://intelli-verse-x.github.io/Intelli-verse-X-Unity-SDK/modules/multiplayer/")]
    public class IVXGameModeManager : MonoBehaviour
    {
        #region Singleton

        private static IVXGameModeManager _instance;

        /// <summary>Singleton accessor. Auto-creates if not present in scene.</summary>
        public static IVXGameModeManager Instance
        {
            get
            {
                if (_instance == null)
                {
                    _instance = FindFirstObjectByType<IVXGameModeManager>();
                    if (_instance == null)
                    {
                        var go = new GameObject("[IVXGameModeManager]");
                        _instance = go.AddComponent<IVXGameModeManager>();
                        DontDestroyOnLoad(go);
                    }
                }
                return _instance;
            }
        }

        #endregion

        #region Events

        /// <summary>Fired when the game mode changes.</summary>
        public event Action<IVXGameMode> OnModeChanged;

        /// <summary>Fired when a player is added to a slot.</summary>
        public event Action<IVXPlayerSlot> OnPlayerAdded;

        /// <summary>Fired when a player is removed from a slot.</summary>
        public event Action<IVXPlayerSlot> OnPlayerRemoved;

        /// <summary>Fired when a player's ready state changes.</summary>
        public event Action<IVXPlayerSlot> OnPlayerReadyChanged;

        /// <summary>Fired when all players are ready and match can start.</summary>
        public event Action<IVXMatchConfig> OnAllPlayersReady;

        /// <summary>Fired when match phase transitions.</summary>
        public event Action<IVXMatchPhase, IVXMatchPhase> OnMatchPhaseChanged;

        /// <summary>Fired when the match config is updated.</summary>
        public event Action<IVXMatchConfig> OnConfigChanged;

        #endregion

        #region Properties

        /// <summary>Current match configuration.</summary>
        public IVXMatchConfig CurrentConfig { get; private set; }

        /// <summary>Currently selected game mode.</summary>
        public IVXGameMode CurrentMode => CurrentConfig?.Mode ?? IVXGameMode.Solo;

        /// <summary>Current match phase.</summary>
        public IVXMatchPhase Phase { get; private set; } = IVXMatchPhase.None;

        /// <summary>All registered player slots.</summary>
        public IReadOnlyList<IVXPlayerSlot> Players => _players;

        /// <summary>Number of occupied slots.</summary>
        public int PlayerCount => _players.Count;

        /// <summary>Whether all minimum players have joined and are ready.</summary>
        public bool CanStartMatch => CurrentConfig != null
            && _players.Count >= CurrentConfig.MinPlayers
            && _players.TrueForAll(p => p.ReadyState == IVXPlayerReadyState.Ready || !p.IsHuman);

        /// <summary>Whether the local player is the host.</summary>
        public bool IsHost { get; private set; } = true;

        /// <summary>Local player slot (always slot 0 for the primary local player).</summary>
        public IVXPlayerSlot LocalPlayer => _players.Count > 0 ? _players[0] : null;

        #endregion

        #region Private Fields

        private readonly List<IVXPlayerSlot> _players = new List<IVXPlayerSlot>();

        #endregion

        #region Unity Lifecycle

        private void Awake()
        {
            if (_instance != null && _instance != this)
            {
                Destroy(gameObject);
                return;
            }
            _instance = this;
            DontDestroyOnLoad(gameObject);
        }

        private void OnDestroy()
        {
            if (_instance == this) _instance = null;
        }

        #endregion

        #region Public Methods — Mode Selection

        /// <summary>Set the game mode and configure default match settings.</summary>
        /// <param name="mode">Desired game mode.</param>
        /// <param name="maxPlayers">Override max players (0 = mode default).</param>
        public void SelectMode(IVXGameMode mode, int maxPlayers = 0)
        {
            IVXMatchConfig config;
            switch (mode)
            {
                case IVXGameMode.Solo:
                    config = IVXMatchConfig.Solo();
                    break;
                case IVXGameMode.LocalMultiplayer:
                    config = IVXMatchConfig.Local(maxPlayers > 0 ? maxPlayers : 4);
                    break;
                case IVXGameMode.OnlineVersus:
                case IVXGameMode.RankedMatch:
                    config = mode == IVXGameMode.RankedMatch
                        ? IVXMatchConfig.Ranked()
                        : IVXMatchConfig.OnlineVersus(maxPlayers > 0 ? maxPlayers : 2);
                    break;
                case IVXGameMode.OnlineCoop:
                    config = IVXMatchConfig.OnlineCoop(maxPlayers > 0 ? maxPlayers : 4);
                    break;
                default:
                    config = new IVXMatchConfig { Mode = mode, MaxPlayers = maxPlayers > 0 ? maxPlayers : 2 };
                    break;
            }

            SetConfig(config);
        }

        /// <summary>Apply a fully custom match configuration.</summary>
        /// <param name="config">The match configuration to use.</param>
        public void SetConfig(IVXMatchConfig config)
        {
            var oldMode = CurrentConfig?.Mode ?? IVXGameMode.Solo;
            CurrentConfig = config ?? throw new ArgumentNullException(nameof(config));

            _players.Clear();

            AddLocalPlayer("Player 1");
            IsHost = true;

            if (config.Mode != oldMode)
                OnModeChanged?.Invoke(config.Mode);

            OnConfigChanged?.Invoke(config);
            SetPhase(IVXMatchPhase.Lobby);

            Debug.Log($"[{nameof(IVXGameModeManager)}] Mode set: {config.Mode}, Max: {config.MaxPlayers}");
        }

        #endregion

        #region Public Methods — Player Management

        /// <summary>Add a local player to the next available slot.</summary>
        /// <param name="displayName">Display name for the player.</param>
        /// <param name="inputDeviceIndex">Input device index (-1 = default).</param>
        /// <returns>The created player slot, or null if full.</returns>
        public IVXPlayerSlot AddLocalPlayer(string displayName, int inputDeviceIndex = -1)
        {
            if (CurrentConfig == null) return null;
            if (_players.Count >= CurrentConfig.MaxPlayers) return null;

            var slot = new IVXPlayerSlot(_players.Count, displayName, true)
            {
                InputDeviceIndex = inputDeviceIndex,
                IsHost = _players.Count == 0
            };

            _players.Add(slot);
            OnPlayerAdded?.Invoke(slot);
            return slot;
        }

        /// <summary>Add a remote (online) player to the next available slot.</summary>
        /// <param name="userId">Remote player user ID.</param>
        /// <param name="displayName">Remote player display name.</param>
        /// <returns>The created player slot, or null if full.</returns>
        public IVXPlayerSlot AddRemotePlayer(string userId, string displayName)
        {
            if (CurrentConfig == null) return null;
            if (_players.Count >= CurrentConfig.MaxPlayers) return null;

            var slot = new IVXPlayerSlot(_players.Count, displayName, false)
            {
                UserId = userId,
                IsHost = false
            };

            _players.Add(slot);
            OnPlayerAdded?.Invoke(slot);
            return slot;
        }

        /// <summary>Add an AI bot to the next available slot.</summary>
        /// <param name="botName">Display name for the bot.</param>
        /// <returns>The created player slot, or null if full or bots not allowed.</returns>
        public IVXPlayerSlot AddBot(string botName)
        {
            if (CurrentConfig == null || !CurrentConfig.AllowBots) return null;
            if (_players.Count >= CurrentConfig.MaxPlayers) return null;

            var slot = new IVXPlayerSlot(_players.Count, botName, true)
            {
                IsHuman = false,
                ReadyState = IVXPlayerReadyState.Ready
            };

            _players.Add(slot);
            OnPlayerAdded?.Invoke(slot);
            return slot;
        }

        /// <summary>Remove a player from a slot.</summary>
        /// <param name="slotIndex">Index of the slot to remove.</param>
        public void RemovePlayer(int slotIndex)
        {
            if (slotIndex < 0 || slotIndex >= _players.Count) return;
            if (slotIndex == 0) return; // Cannot remove primary local player

            var removed = _players[slotIndex];
            _players.RemoveAt(slotIndex);

            for (int i = 0; i < _players.Count; i++)
                _players[i].SlotIndex = i;

            OnPlayerRemoved?.Invoke(removed);
        }

        /// <summary>Set a player's ready state.</summary>
        /// <param name="slotIndex">Player slot index.</param>
        /// <param name="ready">Whether the player is ready.</param>
        public void SetPlayerReady(int slotIndex, bool ready)
        {
            if (slotIndex < 0 || slotIndex >= _players.Count) return;

            _players[slotIndex].ReadyState = ready ? IVXPlayerReadyState.Ready : IVXPlayerReadyState.NotReady;
            OnPlayerReadyChanged?.Invoke(_players[slotIndex]);

            if (CanStartMatch)
                OnAllPlayersReady?.Invoke(CurrentConfig);
        }

        /// <summary>Assign a player to a team.</summary>
        /// <param name="slotIndex">Player slot index.</param>
        /// <param name="team">Team to assign.</param>
        public void SetPlayerTeam(int slotIndex, IVXTeam team)
        {
            if (slotIndex < 0 || slotIndex >= _players.Count) return;
            _players[slotIndex].Team = team;
        }

        #endregion

        #region Public Methods — Match Lifecycle

        /// <summary>Transition the match to a new phase.</summary>
        /// <param name="phase">Target phase.</param>
        public void SetPhase(IVXMatchPhase phase)
        {
            if (Phase == phase) return;
            var old = Phase;
            Phase = phase;
            OnMatchPhaseChanged?.Invoke(old, phase);
            Debug.Log($"[{nameof(IVXGameModeManager)}] Phase: {old} -> {phase}");
        }

        /// <summary>Start the match (transitions to Loading then InProgress).</summary>
        public void StartMatch()
        {
            if (!CanStartMatch)
            {
                Debug.LogWarning($"[{nameof(IVXGameModeManager)}] Cannot start: not enough ready players.");
                return;
            }

            foreach (var p in _players)
                p.ReadyState = IVXPlayerReadyState.Loading;

            SetPhase(IVXMatchPhase.Loading);
        }

        /// <summary>Signal that loading is complete and gameplay begins.</summary>
        public void BeginGameplay()
        {
            foreach (var p in _players)
                p.ReadyState = IVXPlayerReadyState.InGame;

            SetPhase(IVXMatchPhase.InProgress);
        }

        /// <summary>End the current match.</summary>
        public void EndMatch()
        {
            SetPhase(IVXMatchPhase.Finished);
        }

        /// <summary>Reset to lobby state, keeping current config.</summary>
        public void ReturnToLobby()
        {
            foreach (var p in _players)
                p.ReadyState = IVXPlayerReadyState.NotReady;

            SetPhase(IVXMatchPhase.Lobby);
        }

        /// <summary>Full reset — clears config, players, and phase.</summary>
        public void Reset()
        {
            _players.Clear();
            CurrentConfig = null;
            Phase = IVXMatchPhase.None;
            IsHost = true;
        }

        #endregion

        #region Public Methods — Queries

        /// <summary>Get all players on a specific team.</summary>
        /// <param name="team">Team to query.</param>
        /// <returns>List of players on the team.</returns>
        public List<IVXPlayerSlot> GetTeamPlayers(IVXTeam team)
        {
            return _players.FindAll(p => p.Team == team);
        }

        /// <summary>Get all local players (same-device).</summary>
        public List<IVXPlayerSlot> GetLocalPlayers()
        {
            return _players.FindAll(p => p.IsLocal);
        }

        /// <summary>Get all remote (online) players.</summary>
        public List<IVXPlayerSlot> GetRemotePlayers()
        {
            return _players.FindAll(p => !p.IsLocal);
        }

        /// <summary>Check if a specific game mode is available given current SDK configuration.</summary>
        /// <param name="mode">Mode to check.</param>
        /// <returns>True if the mode can be selected.</returns>
        public bool IsModeAvailable(IVXGameMode mode)
        {
            switch (mode)
            {
                case IVXGameMode.Solo:
                case IVXGameMode.LocalMultiplayer:
                case IVXGameMode.TurnBased:
                    return true;
                case IVXGameMode.OnlineMultiplayer:
                case IVXGameMode.OnlineVersus:
                case IVXGameMode.OnlineCoop:
                case IVXGameMode.RankedMatch:
                    return Application.internetReachability != NetworkReachability.NotReachable;
                default:
                    return false;
            }
        }

        #endregion
    }
}

File: SDKs/java/src/main/java/com/intelliversex/sdk/gamemodes/IVXGameModeManager.java
// Copyright (c) 2026 Intelli-verse-X
// MIT License — see LICENSE in the project root.

package com.intelliversex.sdk.gamemodes;

import com.intelliversex.sdk.gamemodes.IVXGameModeModels.IVXMatchResult;
import com.intelliversex.sdk.gamemodes.IVXGameModeModels.IVXPlayerSlot;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

/**
 * Thread-safe singleton that manages game-mode selection, player slots,
 * readiness checks, and match lifecycle.
 * <p>
 * Register callbacks via the {@code onXxx} methods to react to state changes.
 * <p>
 * <b>Usage:</b>
 * <pre>{@code
 * IVXGameModeManager mgr = IVXGameModeManager.getInstance();
 * mgr.selectMode(IVXGameMode.ONLINE_VERSUS, 4);
 * mgr.addPlayer("Alice", true);
 * mgr.setPlayerReady(0, true);
 * if (mgr.canStartMatch()) mgr.startMatch();
 * }</pre>
 */
public class IVXGameModeManager {

    /**
     * Supported game modes.
     */
    public enum IVXGameMode {
        SOLO,
        LOCAL_MULTIPLAYER,
        ONLINE_VERSUS,
        ONLINE_COOP,
        RANKED,
        TURN_BASED
    }

    private static volatile IVXGameModeManager instance;

    private volatile IVXGameMode currentMode;
    private volatile int maxPlayers;
    private volatile boolean matchActive;
    private final List<IVXPlayerSlot> players = new CopyOnWriteArrayList<>();

    private final List<Consumer<IVXGameMode>> modeChangedListeners = new CopyOnWriteArrayList<>();
    private final List<Consumer<IVXPlayerSlot>> playerAddedListeners = new CopyOnWriteArrayList<>();
    private final List<Consumer<Integer>> playerRemovedListeners = new CopyOnWriteArrayList<>();
    private final List<Consumer<IVXPlayerSlot>> playerReadyListeners = new CopyOnWriteArrayList<>();
    private final List<Consumer<Boolean>> matchStateListeners = new CopyOnWriteArrayList<>();

    private IVXGameModeManager() {}

    /**
     * Returns the singleton instance, creating it on first access.
     *
     * @return the shared {@link IVXGameModeManager} instance
     */
    public static IVXGameModeManager getInstance() {
        if (instance == null) {
            synchronized (IVXGameModeManager.class) {
                if (instance == null) {
                    instance = new IVXGameModeManager();
                }
            }
        }
        return instance;
    }

    // ──────────────────────────────────────────────
    //  Mode Selection
    // ──────────────────────────────────────────────

    /**
     * Selects a game mode and sets the maximum number of players.
     * Resets any existing player slots.
     *
     * @param mode       the game mode to use
     * @param maxPlayers maximum players allowed (must be &ge; 1)
     * @throws IllegalArgumentException if maxPlayers &lt; 1
     */
    public synchronized void selectMode(IVXGameMode mode, int maxPlayers) {
        if (mode == null) {
            throw new IllegalArgumentException("mode must not be null");
        }
        if (maxPlayers < 1) {
            throw new IllegalArgumentException("maxPlayers must be >= 1, got: " + maxPlayers);
        }
        this.currentMode = mode;
        this.maxPlayers = maxPlayers;
        this.matchActive = false;
        this.players.clear();
        modeChangedListeners.forEach(l -> l.accept(mode));
    }

    /**
     * Returns the currently selected game mode, or {@code null} if none selected.
     *
     * @return the current {@link IVXGameMode}
     */
    public IVXGameMode getCurrentMode() { return currentMode; }

    /**
     * Returns the maximum number of players for the current mode.
     *
     * @return max player count
     */
    public int getMaxPlayers() { return maxPlayers; }

    // ──────────────────────────────────────────────
    //  Player Management
    // ──────────────────────────────────────────────

    /**
     * Adds a player to the next available slot.
     *
     * @param name    display name for the player
     * @param isLocal {@code true} if this is a local player
     * @return the newly created {@link IVXPlayerSlot}
     * @throws IllegalStateException if the lobby is full or no mode is selected
     */
    public synchronized IVXPlayerSlot addPlayer(String name, boolean isLocal) {
        requireModeSelected();
        if (players.size() >= maxPlayers) {
            throw new IllegalStateException("Lobby is full (" + maxPlayers + " players max)");
        }
        IVXPlayerSlot slot = new IVXPlayerSlot(players.size(), name, isLocal);
        players.add(slot);
        playerAddedListeners.forEach(l -> l.accept(slot));
        return slot;
    }

    /**
     * Removes the player at the given slot index.
     *
     * @param slotIndex the slot to remove
     * @throws IndexOutOfBoundsException if the index is invalid
     */
    public synchronized void removePlayer(int slotIndex) {
        requireModeSelected();
        if (slotIndex < 0 || slotIndex >= players.size()) {
            throw new IndexOutOfBoundsException("Invalid slot index: " + slotIndex);
        }
        players.remove(slotIndex);
        playerRemovedListeners.forEach(l -> l.accept(slotIndex));
    }

    /**
     * Sets the ready state for a player slot.
     *
     * @param slot  the slot index
     * @param ready {@code true} to mark ready
     * @throws IndexOutOfBoundsException if the index is invalid
     */
    public synchronized void setPlayerReady(int slot, boolean ready) {
        requireModeSelected();
        if (slot < 0 || slot >= players.size()) {
            throw new IndexOutOfBoundsException("Invalid slot index: " + slot);
        }
        IVXPlayerSlot ps = players.get(slot);
        ps.setReady(ready);
        playerReadyListeners.forEach(l -> l.accept(ps));
    }

    /**
     * Returns an unmodifiable snapshot of the current player list.
     *
     * @return list of {@link IVXPlayerSlot}
     */
    public List<IVXPlayerSlot> getPlayers() {
        return Collections.unmodifiableList(new ArrayList<>(players));
    }

    // ──────────────────────────────────────────────
    //  Match Lifecycle
    // ──────────────────────────────────────────────

    /**
     * Returns {@code true} if the match can start: a mode is selected,
     * at least one player is present, and all players are ready.
     *
     * @return whether the match is ready to begin
     */
    public boolean canStartMatch() {
        if (currentMode == null || players.isEmpty()) return false;
        for (IVXPlayerSlot p : players) {
            if (!p.isReady()) return false;
        }
        return true;
    }

    /**
     * Starts the match. Requires {@link #canStartMatch()} to return {@code true}.
     *
     * @throws IllegalStateException if preconditions are not met
     */
    public synchronized void startMatch() {
        if (!canStartMatch()) {
            throw new IllegalStateException("Cannot start match — check canStartMatch()");
        }
        this.matchActive = true;
        matchStateListeners.forEach(l -> l.accept(true));
    }

    /**
     * Ends the current match.
     *
     * @throws IllegalStateException if no match is active
     */
    public synchronized void endMatch() {
        if (!matchActive) {
            throw new IllegalStateException("No active match to end");
        }
        this.matchActive = false;
        matchStateListeners.forEach(l -> l.accept(false));
    }

    /**
     * Returns {@code true} if a match is currently in progress.
     *
     * @return match active state
     */
    public boolean isMatchActive() { return matchActive; }

    /**
     * Resets all state: clears mode, players, and match flag.
     */
    public synchronized void reset() {
        this.currentMode = null;
        this.maxPlayers = 0;
        this.matchActive = false;
        this.players.clear();
    }

    // ──────────────────────────────────────────────
    //  Event Callbacks
    // ──────────────────────────────────────────────

    /**
     * Registers a callback fired when the game mode changes.
     *
     * @param listener the callback
     */
    public void onModeChanged(Consumer<IVXGameMode> listener) {
        modeChangedListeners.add(listener);
    }

    /**
     * Registers a callback fired when a player is added.
     *
     * @param listener the callback
     */
    public void onPlayerAdded(Consumer<IVXPlayerSlot> listener) {
        playerAddedListeners.add(listener);
    }

    /**
     * Registers a callback fired when a player is removed (receives slot index).
     *
     * @param listener the callback
     */
    public void onPlayerRemoved(Consumer<Integer> listener) {
        playerRemovedListeners.add(listener);
    }

    /**
     * Registers a callback fired when a player's ready state changes.
     *
     * @param listener the callback
     */
    public void onPlayerReady(Consumer<IVXPlayerSlot> listener) {
        playerReadyListeners.add(listener);
    }

    /**
     * Registers a callback fired when the match starts ({@code true}) or ends ({@code false}).
     *
     * @param listener the callback
     */
    public void onMatchStateChanged(Consumer<Boolean> listener) {
        matchStateListeners.add(listener);
    }

    // ──────────────────────────────────────────────
    //  Internals
    // ──────────────────────────────────────────────

    private void requireModeSelected() {
        if (currentMode == null) {
            throw new IllegalStateException("No game mode selected. Call selectMode() first.");
        }
    }
}
